### DISCUSSION TITLE
Kitten to paste last command output into current command prompt

### DISCUSSION THREAD

**Discussion by fmagin** [kovidgoyal/kitty#4477] [Q&A]
I really like the new shell integration features in 0.24 and I am currently trying to write a small helper/kitten for something I wanted for a long time: Paste the result of the last command at the cursor position:

```console
fmagin (calibre) ~> which python
/home/fmagin/.virtualenvs/calibre/bin/python
fmagin@monolith ~> readlink -f <want to insert the result of `which python` here> 
```
The [docs for kittens](https://sw.kovidgoyal.net/kitty/kittens/custom/#passing-the-contents-of-the-screen-to-the-kitten) list the various kinds of `type_of_input` I can pass to the `result_handler`, but it is not clear to me where this then available. 

`no_ui` means the `main` function isn't executed, but the `main` function had the comment:

> in main, STDIN is for the kitten process and will contain the contents of the screen

This sounds like the `last_visited_output` I want is somehow only available in the `main` function, but I would prefer to keep `no_ui=True` to prevent the somewhat annoying flicker.

```python
def main(args):
    return "main_result"

# in handle_result, STDIN is for the kitty process itself, rather
# than the kitten process and should not be read from.
from kittens.tui.handler import result_handler
@result_handler(type_of_input='last_visited_output',no_ui=True)
def handle_result(args, stdin_data, target_window_id, boss):
    w = boss.window_id_map.get(target_window_id)
    if w is not None:
        w.paste(stdin_data)
```

with `no_ui=False` this at least pastes `main_result`. But `def main(args): return sys.stdin.read()` pastes nothing again.

What I had expected and found ideal for my scenario would be that with `no_ui=True` the requested `type_of_input` is directly available in `stdin_data` and I could simply use it for pasting.
Where in this process am I misunderstanding the design and execution of kittens?

**Comment by page-down** [comment]
> But def main(args): return sys.stdin.read() pastes nothing again.

You need `output`, and `last_visited_output` needs to be `visited` by keyboard (ctrl+shift+z/x) or mouse (ctrl+shift+right-click) to be valid.

> ... only available in the main function, ... prefer to keep no_ui=True to prevent the somewhat annoying flicker.

I'm also curious to know how this can be done, which seems unlikely at the moment.

Here is the source code that affects this:

kitty/boss.py
```python
def _run_kitten( ... ) -> Any:
    # ...
    if end_kitten.no_ui:
        return end_kitten(None, getattr(w, 'id', None), self)
    # ...
```

You can see that stdin_data is given None and then returned directly.

EDIT:
Here is a working example. I really don't like this approach.

```python
from typing import List

from kitty.boss import Boss
from kittens.tui.handler import result_handler


def main(args: List[str]) -> None:
    return None


@result_handler(no_ui=True)
def handle_result(args: List[str], stdin_data: str, target_window_id: int, boss: Boss) -> None:
    w = boss.window_id_map.get(target_window_id)
    if w is not None and w.screen.cursor_at_prompt():
        w.paste(w.cmd_output())
```

**Reply by fmagin** [reply]
I assumed `last_visited_output` would be the last output by default. Is this incorrect?

I like your working example though, this definitely covers most of my use cases already, choosing another command by using `last_visited_output` was more of a bonus anyway.

This is somewhat tangential, but is there some way to run the kitten from a fish script, so I can then e.g. pipe it into `fzf` (and then paste the result of that, but that pasting I can easily achieve in `fish`)?
Adding a `print` to `handle_result` doesn't give me any output if I run `kitty +kitten get_result.py`

**Reply by page-down** [reply]
> I assumed last_visited_output would be the last output by default. Is this incorrect?

This is empty by default.

https://sw.kovidgoyal.net/kitty/kittens/custom/#passing-the-contents-of-the-screen-to-the-kitten

> ... run the kitten from a fish script, so I can then e.g. pipe it into fzf (and then paste the result of that ...

It sounds like you need remote control. If you want to operate at the shell prompt, then fzf obviously needs to be run by the shell. Get the text and send it to fzf for further processing.

https://sw.kovidgoyal.net/kitty/remote-control/#kitty-get-text

> ... doesn't give me any output if I run kitty +kitten get_result.py

Running kitten from the command line with kitty will only involve the main() part. This is a standalone cli process and is not executed in the runtime of the kitty GUI program.

**Comment by permelin** [comment]
I just learned that you can get the effect you want with a mapping like this:
```
map f1 combine : launch --stdin-source=@last_cmd_output --type=primary : paste_from_selection
```
It copies the output from the last command to the primary selection and then pastes it.
